src/libraries/Microsoft.Extensions.Diagnostics/src/Metrics/ListenerSubscription.cs

```diff
@@ -0,0 +1,324 @@
+        {
+            var alreadyEnabled = status.Enabled;
+            var enable = false;
+            var rule = GetMostSpecificRule(instrument);
+            if (rule != null)
+            {
+                enable = rule.Enable;
+            }
+
+            if (!enable && alreadyEnabled)
+            {
+                status.Enabled = false;
+                _meterListener.DisableMeasurementEvents(instrument);
+            }
+            else if (enable && !alreadyEnabled)
+            {
+                // The first time we enable an instrument, we need to call InstrumentPublished.
+                if (!status.Published)
+                {
+                    // However, a listener might decline to enable the instrument, remember that.
+                    status.Published = true;
+                    status.ListenerEnabled = _metricsListener.InstrumentPublished(instrument, out var state);
+                    status.State = state;
+                }
+
+                if (status.ListenerEnabled)
+                {
+                    _meterListener.EnableMeasurementEvents(instrument, status.State);
+                    status.Enabled = true;
+                }
+            }
+        }
+
+        private InstrumentRule? GetMostSpecificRule(Instrument instrument)
+        {
+            InstrumentRule? best = null;
+            foreach (var rule in _rules)
+            {
+                if (RuleMatches(rule, instrument, _metricsListener.Name)
+                    && IsMoreSpecific(rule, best, isLocalScope: instrument.Meter.Scope != null))
+                {
+                    best = rule;
+                }
+            }
+
+            return best;
+        }
+
+        // internal for testing
+        internal static bool RuleMatches(InstrumentRule rule, Instrument instrument, string listenerName)
+        {
+            // Exact match or empty
+            if (!string.IsNullOrEmpty(rule.ListenerName)
+                && !string.Equals(rule.ListenerName, listenerName, StringComparison.OrdinalIgnoreCase))
+            {
+                return false;
+            }
+
+            // Exact match or empty
+            if (!string.IsNullOrEmpty(rule.InstrumentName)
+                && !string.Equals(rule.InstrumentName, instrument.Name, StringComparison.OrdinalIgnoreCase))
+            {
+                return false;
+            }
+
+            if (!(rule.Scopes.HasFlag(MeterScope.Global) && instrument.Meter.Scope == null)
+                && !(rule.Scopes.HasFlag(MeterScope.Local) && instrument.Meter.Scope != null)) // TODO: What should we be comparing Scope to, the DefaultMeterFactory / IMeterFactory?
+            {
+                return false;
+            }
+
+            // Meter
+
+            var ruleMeterName = rule.MeterName.AsSpan();
+            // Rule "System.Net.*" matches meter "System.Net" and "System.Net.Http"
+            if (ruleMeterName.EndsWith(".*".AsSpan(), StringComparison.Ordinal))
```

Comment:
>if (ruleMeterName.EndsWith(".*".AsSpan(), StringComparison.Ordinal)) [](http://example.com/codeflow?start=12&length=68)

what happen if someone specify the rule `.*` only? would it be ok to have this equal to the rule with empty string or `*` only? I think this may be confusing.

Reply:
Being discussed over at https://github.com/dotnet/runtime/pull/90201#discussion_r1289356270